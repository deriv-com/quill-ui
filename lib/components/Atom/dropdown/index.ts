export * from "./item";
export * from "./divider";
export * from "./title";
export * from "./item-container";
