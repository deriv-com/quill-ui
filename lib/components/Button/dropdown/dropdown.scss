.dropdown-menu {
    &__box{
        padding-bottom:  var(--core-spacing-200);
    }
    &__container{
        background-color: var(--component-dropdownList-bg);
        border-radius: var(--core-borderRadius-400);
        box-shadow: var(--semantic-elevation-shadow-card-hover);
        display: flex;
        padding: var(--core-spacing-400);
        flex-direction: column;
        align-items: flex-start;
        max-height: 258px;
        width: fit-content;
        overflow-y: auto;
    }
    &__item{
        display: flex;
        padding: var(--core-spacing-50) var(--core-spacing-600);
        align-items: center;
        gap: var(--core-spacing-400);
        align-self: stretch;
        border-radius: var(--core-borderRadius-200);
        color: var(--component-dropdownItem-label-color-default);

        &--size {
           &-sm {
            min-height: var(--core-borderRadius-1600);
           }
           &-md {
            min-height: var(--core-borderRadius-1600);
           }
           &-lg {
            min-height: var(--core-borderRadius-2400);
           }
           &-xl {
            min-height: var(--core-borderRadius-2400);
          }
        }

        &:hover{
            background-color: var(--component-dropdownItem-bg-hover);
            cursor: pointer ;
      
        }
        &--selected:not(:active):not(hover){
            background-color: var(--component-dropdownItem-bg-selected);
            color : var(--component-dropdownItem-label-color-selectedWhite);
        }
        &--active{
            background-color: var(--component-dropdownItem-bg-active);
        
        }
    }
    &__transition {
        &--enter {
            transform-origin: top;
            animation-timing-function: var(--core-motion-ease-400);
            animation-duration: var(--core-motion-duration-200);
        }

    }
}