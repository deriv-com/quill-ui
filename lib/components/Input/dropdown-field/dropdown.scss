.dropdown {
    &__wrapper {
        position: relative;
        max-width: 360px;

        ul {
            margin: var(--core-spacing-50);
            padding: var(--core-spacing-50);
        }
    }

    &__container {
        position: absolute;
        background-color: var(--component-dropdownList-bg);
        border-radius: var(--core-borderRadius-400);
        box-shadow: var(--semantic-elevation-shadow-card-hover);
        display: flex;
        flex-direction: column;
        align-items: flex-start;
        margin-top: var(--core-size-200);
        min-width: 360px;
        overflow-y: auto;

        &--height {
            max-height: 420px;
        }

        &--size {
            &-sm {
                padding: var(--core-spacing-200);
            }

            &-md {
                padding: var(--core-spacing-400);
            }
        }
    }

    &__item {
        &--active {
            background-color: var(--component-dropdownItem-bg-active);
        }
    }
}
.deriv-dropdown__menu {
    transform-origin: top;
    transform: scaleY(0);
    transition: transform var(--core-motion-duration-200)
        var(--core-motion-ease-400);

    &--open {
        opacity: 1;
        transform: scaleY(1);
    }

    &--close {
        opacity: 0;
        transform: scaleY(0);
    }
}
