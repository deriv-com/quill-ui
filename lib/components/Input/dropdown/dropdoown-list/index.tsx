import { Listbox } from "@headlessui/react";
import React, { Fragment } from "react";
import DropdownItem, { DropdownItemProps } from "../dropdown-item";
import clsx from "clsx";
import { SingleSelectChipProps } from "@components/Button";

export type TSingleSelectItem = {
    value: number | string;
    label: React.ReactNode;
    disabled?: boolean;
};

export interface OptionProps extends DropdownItemProps {
    item: TSingleSelectItem;
}

const Option = ({ item, centered }: OptionProps) => {
    return (
        <Listbox.Option value={item} as={Fragment} disabled={item.disabled}>
            {({ disabled, selected }) => (
                <DropdownItem
                    label={item.label}
                    selected={selected}
                    disabled={disabled}
                    centered={centered}
                />
            )}
        </Listbox.Option>
    );
};

export const DropdownList = ({ ...rest }: SingleSelectChipProps) => {
    return (
        <Listbox.Options
            className={clsx("dropdown", `dropdown__size--${size}`, className)}
        >
            <Option item={defaultOption} size={size} />
            {options.map((item) => (
                <Option
                    item={item}
                    key={`chip-selectable-item-${item.value}`}
                    size={size}
                />
            ))}
        </Listbox.Options>
    );
};

export default DropdownList;
