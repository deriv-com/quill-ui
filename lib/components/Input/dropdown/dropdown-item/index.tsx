import React, { ReactNode } from "react";
import { TMediumSizes } from "@types";
import "./dropdown-item.scss";
import clsx from "clsx";
import { Text } from "@components/Typography";

export interface DropdownItemProps {
    leftIcon?: ReactNode;
    rightIcon?: ReactNode;
    as?: "div" | "li";
    size?: TMediumSizes;
    centered?: boolean;
    label: ReactNode;
    disabled?: boolean;
    selected?: boolean;
    className?: string;
}

export const DropdownItem = ({
    as: Component = "div",
    leftIcon,
    rightIcon,
    size = "md",
    centered = false,
    label,
    disabled = false,
    selected = false,
    className,
}: DropdownItemProps) => {
    return (
        <Component
            className={clsx(
                "quill-dropdown-item",
                `quill-dropdown-item__size--${size}`,
                `quill-dropdown-item__selected--${selected}__disabled--${disabled}`,
                className,
            )}
        >
            {leftIcon}
            <Text
                size={size}
                color="disabled"
                className={clsx(
                    "quill-dropdown-item-label",
                    disabled && "quill-typography__color--disabled",
                )}
                centered={centered}
            >
                {label}
            </Text>
            {rightIcon}
        </Component>
    );
};

export default DropdownItem;
