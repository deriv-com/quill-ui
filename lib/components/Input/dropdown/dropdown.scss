.dropdown {
    &__container {
        background-color: var(--component-dropdownList-bg);
        border-radius: var(--core-borderRadius-400);
        box-shadow: var(--semantic-elevation-shadow-card-hover);
        display: flex;
        padding: var(--core-spacing-400);
        flex-direction: column;
        align-items: flex-start;
        max-height: 420px;
        margin-top: 8px;
        max-width: 360px;
        overflow-y: auto;
    }
    &__item {
        cursor: pointer;
        display: flex;
        padding: var(--core-spacing-50) var(--core-spacing-600);
        align-items: center;
        width: 100%;
        gap: var(--core-spacing-400);
        align-self: stretch;
        border-radius: var(--core-borderRadius-200);
        color: var(--component-dropdownItem-label-color-default);
        &--size {
            &-sm {
               height: 32px;
            }
            &-md {
               height: 48px;
            }
         
        }
        &:hover {
            background-color: var(--component-dropdownItem-bg-hover);
            cursor: pointer;
        }
        &--selected:not(:active):not(hover) {
            background-color: var(--component-dropdownItem-bg-selected);
            color: var(--component-dropdownItem-label-color-selectedWhite);
        }
        &--active {
            background-color: var(--component-dropdownItem-bg-active);
        }
    }
    &__icon {
        &-svg {
            fill: var(--component-textIcon-normal-prominent);
        }
    }
}