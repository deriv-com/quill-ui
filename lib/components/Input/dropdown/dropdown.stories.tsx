import { Meta, StoryObj } from "@storybook/react";
import DropdownFeild from ".";
import {
    StandaloneCircleCheckBoldIcon,
    StandalonePlaceholderRegularIcon,
    StandaloneTriangleExclamationBoldIcon,
} from "@deriv/quill-icons";
import { LabelPairedPlaceholderSmRegularIcon } from "@deriv/quill-icons/LabelPaired";

const icons: Record<string, object | null> = {
    with_icon: <StandalonePlaceholderRegularIcon iconSize="sm" />,
    none: null,
};

const statusIcon: Record<string, object | null> = {
    placeholder: <StandalonePlaceholderRegularIcon iconSize="sm" />,
    success: <StandaloneCircleCheckBoldIcon iconSize="sm" />,
    error: <StandaloneTriangleExclamationBoldIcon iconSize="sm" />,
    none: null,
};

const placeholder = "Select";

const meta = {
    title: "Components/Inputs/Dropdown",
    component: DropdownFeild,
    parameters: {
        docs: {
            story: {
                height: "350px",
                width: "350px",
            },
        },
    },
    tags: ["autodocs"],
    args: {
        inputSize: "md",
        status: "neutral",
        disabled: false,
        variant: "fill",
        textAlignment: "left",
        fieldMarker: false,
        required: false,
    },
    argTypes: {
        inputSize: {
            control: {
                type: "radio",
            },
            options: ["sm", "md"],
        },
        leftIcon: {
            options: Object.keys(icons),
            mapping: icons,
            control: {
                type: "select",
            },
        },
        rightIcon: {
            options: Object.keys(statusIcon),
            mapping: statusIcon,
            control: {
                type: "select",
            },
        },
        status: {
            control: {
                type: "radio",
            },
            options: ["neutral", "success", "error"],
        },
        variant: {
            control: {
                type: "radio",
            },
            options: ["fill", "outline"],
        },
        label: {
            control: {
                type: "text",
            },
        },
        textAlignment: {
            control: {
                type: "radio",
            },
            options: ["left", "center"],
        },
        leftStatusMessage: {
            control: {
                type: "text",
            },
        },
        rightStatusMessage: {
            control: {
                type: "text",
            },
        },
    },
} satisfies Meta<typeof DropdownFeild>;

export default meta;

type Story = StoryObj<typeof meta>;

export const DropdownFeildWithoutLabel: Story = {
    args: {
        placeholder,
        leftIcon: <LabelPairedPlaceholderSmRegularIcon />,
    },
};
