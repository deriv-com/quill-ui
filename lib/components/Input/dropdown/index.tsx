import React, { Fragment, forwardRef, useState } from "react";
import { Combobox } from "@headlessui/react";
import Input from "../base";
import { TextFieldProps } from "../text-field";
import { LabelPairedChevronDownSmBoldIcon } from "@deriv/quill-icons/LabelPaired";
import clsx from "clsx";
import { Text } from "@components/Typography";
import { TMediumSizes } from "@types";
import "./dropdown.scss";

const people = [
    "Durward Reynolds",
    "Kenton Towne",
    "Therese Wunsch",
    "Benedict Kessler",
    "Katelyn Rohan",
];
export type TSingleSelectItem = {
    value: number | string;
    label: string | React.ReactNode;
};
export interface DropdownOptionProps {
    item: string;
    closeDropdown: () => void;
    inputSize?: TMediumSizes;
    handleKeyDown: (e: React.KeyboardEvent) => void;
}

const Options = ({
    item,
    inputSize,
    closeDropdown,
    handleKeyDown,
}: DropdownOptionProps) => {
    return (
        <Combobox.Option value={item} as={Fragment}>
            {({ selected, active }) => {
                return (
                    <Combobox.Button
                        onKeyDown={handleKeyDown}
                        onClick={closeDropdown}
                        className={clsx(
                            "dropdown__item",
                            `dropdown__item--size-${inputSize}`,
                            selected && "dropdown__item--selected",
                            active && `dropdown__item--active`,
                        )}
                    >
                        <Text
                            as="span"
                            color="var(--component-dropdownItem-label-color-selectedWhite)"
                        >
                            {item}
                        </Text>
                    </Combobox.Button>
                );
            }}
        </Combobox.Option>
    );
};

export const InputDropdown = forwardRef<HTMLInputElement, TextFieldProps>(
    ({ inputSize = "md", ...rest }, ref) => {
        const [isDropdownOpen, setDropdownOpen] = useState(false);
        const handleDropdownClick = () => {
            setDropdownOpen(!isDropdownOpen);
        };

        const closeDropdown = () => {
            setDropdownOpen(false);
        };
        const handleKeyDown = (event: { key: string }) => {
            if (event.key === "Enter") {
                setDropdownOpen(false);
            }
        };

        return (
            <Combobox>
                <Combobox.Input
                    as={Input}
                    type="select"
                    readOnly
                    inputSize={inputSize}
                    onClick={handleDropdownClick}
                    onKeyDown={handleKeyDown}
                    {...rest}
                    rightIcon={
                        <LabelPairedChevronDownSmBoldIcon
                            data-state={isDropdownOpen ? "open" : "close"}
                            className={clsx(
                                "quill-button__transform",
                                isDropdownOpen &&
                                    "quill-button__transform-rotate",
                            )}
                        />
                    }
                    ref={ref}
                />

                {isDropdownOpen && (
                    <div className="dropdown__container">
                        {people.map((person) => (
                            <Options
                                item={person}
                                key={person}
                                closeDropdown={closeDropdown}
                                handleKeyDown={handleKeyDown}
                                inputSize={inputSize}
                            />
                        ))}
                    </div>
                )}
            </Combobox>
        );
    },
);

export default InputDropdown;
