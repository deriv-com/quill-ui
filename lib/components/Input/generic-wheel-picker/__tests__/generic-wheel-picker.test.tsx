import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { GenericWheelPicker } from "..";
import { DropdownProvider } from "@providers/dropdown/dropdownProvider";
import { useDropdown } from "@hooks/useDropdown";
import useBreakpoints from "@hooks/useBreakpoints";
import { act } from "react-dom/test-utils";

const mockData = [
    [
        { value: "1" },
        { value: "2" },
        { value: "3" },
    ],
    [
        { value: "4" },
        { value: "5" },
        { value: "6" },
    ],
    [
        { value: "7"},
        { value: "8"},
        { value: "9"},
    ],
];

jest.mock("@hooks/useDropdown", () => ({
    useDropdown: jest.fn(),
}));
jest.mock("@hooks/useBreakpoints");

class ResizeObserver {
    observe() {}
    unobserve() {}
    disconnect() {}
}

global.ResizeObserver = ResizeObserver;
const mockOnValueChange = jest.fn();

const renderComponent = () =>
    render(
        <DropdownProvider>
            <GenericWheelPicker
                data={mockData}
                values={["1", "4", "7"]}
                label="test"
                onValueChange={mockOnValueChange}
            />
        </DropdownProvider>,
    );

describe("GenericWheelPicker", () => {
    const mockOpen = jest.fn();
    const mockClose = jest.fn();

    beforeEach(() => {
        jest.clearAllMocks();
        (useDropdown as jest.Mock).mockReturnValue({
            isOpen: false,
            open: mockOpen,
            close: mockClose,
        });
        (useBreakpoints as jest.Mock).mockReturnValue({ isMobile: false });
    });

    it("renders correctly", () => {
        const { container } = renderComponent();

        expect(screen.getByRole("textbox")).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    it("should open dropdown on click", async () => {
        renderComponent();

        const input = screen.getByRole("textbox");
        await userEvent.click(input);

        expect(mockOpen).toHaveBeenCalled();
    });

    it("should handle ArrowDown keydown event", async () => {
        renderComponent();

        const input = screen.getByRole("textbox");
        await userEvent.type(input, "{arrowdown}");

        expect(mockOpen).toHaveBeenCalled();
    });

    it("should call onValueChange when dataValues change", async () => {
        renderComponent();

        const input = screen.getByRole("textbox");
        await userEvent.type(input, "{arrowdown}");

        expect(mockOnValueChange).toHaveBeenCalledWith(["1", "4", "7"]);
    });

    it("should render ActionSheet in mobile view", async () => {
        (useBreakpoints as jest.Mock).mockReturnValue({ isMobile: true });
        (useDropdown as jest.Mock).mockReturnValue({
            isOpen: true,
            open: mockOpen,
            close: mockClose,
        })

        renderComponent();

        const input = screen.getByRole("textbox");
        await act(() => {
            userEvent.click(input);
        });
        expect(screen.getByText("6")).toBeInTheDocument();
    });
});
