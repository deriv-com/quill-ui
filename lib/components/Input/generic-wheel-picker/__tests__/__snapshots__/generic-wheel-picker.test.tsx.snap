// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenericWheelPicker renders correctly 1`] = `
<div>
  <div
    class="quill-generic-picker__container quill-generic-picker__is-open--false"
  >
    <div>
      <div
        class="quill-input__container"
      >
        <div
          class="quill-input__wrapper quill-input__wrapper__variant--outline quill-input__wrapper__variant--outline--neutral quill-input__wrapper__size--md quill-generic-picker__input quill-generic-picker__input--hasValue--false"
        >
          <div
            class="quill-input-icons__wrapper"
          >
            <div
              class="quill-input-label__wrapper"
            >
              <input
                autocomplete="off"
                class="input peer input__align--left input__size--md"
                data="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
                id="quill-input-1"
                placeholder=""
                type="text"
                value=""
              />
              <label
                class="label label__status--neutral"
                for="quill-input-1"
              >
                test
              </label>
            </div>
            <svg
              class="quill-input__rotate quill-input__icon"
              data-state="close"
              height="24"
              role="img"
              viewBox="0 0 14 22"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g>
                <path
                  d="m6.535 14.715-5.25-5.25c-.273-.246-.273-.656 0-.93.246-.246.656-.246.93 0L7 13.348l4.785-4.786c.246-.273.656-.273.93 0a.644.644 0 0 1 0 .903l-5.277 5.25a.6.6 0 0 1-.903 0"
                />
              </g>
              <defs>
                <clippath
                  id="b2fbae3ff4787741870694db3344c0c7__a"
                >
                  <path
                    d="M0 0h14v22H0z"
                  />
                </clippath>
              </defs>
            </svg>
          </div>
        </div>
      </div>
    </div>
    <div
      class="quill-generic-picker__content--container"
    />
  </div>
</div>
`;
